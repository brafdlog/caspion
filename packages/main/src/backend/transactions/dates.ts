export const compareObjectsByDate = (a: { date: string }, b: { date: string }) =>
  compareDateStrings(a.date, b.date);
const compareDateStrings = (a: string, b: string) => compareDates(new Date(a), new Date(b));
const compareDates = (a: Date, b: Date) => a.getTime() - b.getTime();
