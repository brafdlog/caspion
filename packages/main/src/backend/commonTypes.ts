import { type CompanyTypes, type ScraperCredentials } from 'israeli-bank-scrapers-core';
import { type Transaction } from 'israeli-bank-scrapers-core/lib/transactions';
import { type Account, type BudgetSummary } from 'ynab';
import { type EventPublisher } from './eventEmitters/EventEmitter';
import { type Credentials } from './export/outputVendors/googleSheets/googleSheetsInternalAPI';

export type { ScraperScrapingResult } from 'israeli-bank-scrapers-core';
export type {
  Credentials,
  OAuth2Client,
  Spreadsheet,
} from './export/outputVendors/googleSheets/googleSheetsInternalAPI';
export interface Config {
  outputVendors: {
    [OutputVendorName.GOOGLE_SHEETS]?: GoogleSheetsConfig;
    [OutputVendorName.YNAB]?: YnabConfig;
    [OutputVendorName.JSON]?: JsonConfig;
    [OutputVendorName.CSV]?: CsvConfig;
  };
  scraping: {
    numDaysBack: number;
    showBrowser: boolean;
    accountsToScrape: AccountToScrapeConfig[];
    chromiumPath?: string;
    maxConcurrency?: number;
    timeout: number;
  };
  useReactUI?: boolean;
}

export enum OutputVendorName {
  YNAB = 'ynab',
  GOOGLE_SHEETS = 'googleSheets',
  JSON = 'json',
  CSV = 'csv',
}

export type OutputVendorConfigs = Exclude<Config['outputVendors'][OutputVendorName], undefined>;
export type OutputVendorConfig<T extends OutputVendorName> = Exclude<
  Config['outputVendors'][T],
  undefined
>;

interface OutputVendorConfigBase {
  active: boolean;
}

export interface CsvConfig extends OutputVendorConfigBase {
  options: {
    filePath: string;
  };
}

export interface JsonConfig extends OutputVendorConfigBase {
  options: {
    filePath: string;
  };
}

export interface GoogleSheetsConfig extends OutputVendorConfigBase {
  options: {
    credentials: Credentials;
    spreadsheetId: string;
  };
}

export interface YnabConfig extends OutputVendorConfigBase {
  options: {
    accessToken: string;
    accountNumbersToYnabAccountIds: Record<string, string>;
    budgetId: string;
    maxPayeeNameLength?: number;
  };
}

export interface AccountToScrapeConfig {
  id: string;
  key: CompanyTypes;
  name: string;
  loginFields: ScraperCredentials;
  active?: boolean;
}

export interface EnrichedTransaction extends Transaction {
  accountNumber: string;
  category?: string;
  hash: string;
}

export interface ExportTransactionsParams {
  transactionsToCreate: EnrichedTransaction[];
  startDate: Date;
  outputVendorsConfig: Config['outputVendors'];
}

export interface ExportTransactionsResult {
  exportedTransactionsNum: number;
}

export type ExportTransactionsFunction = (
  exportTransactionsParams: ExportTransactionsParams,
  eventPublisher: EventPublisher,
) => Promise<ExportTransactionsResult>;

export interface OutputVendor {
  name: OutputVendorName;
  init?: (outputVendorsConfig: Config['outputVendors']) => Promise<void>;
  exportTransactions: ExportTransactionsFunction;
}

export interface FinancialAccountDetails {
  name: string;
  accountNumber: string;
}

export type YnabFinancialAccount = Pick<Account, 'id' | 'name' | 'type'> & {
  budgetId: string;
  active: boolean;
};

export enum FETCH_YNAB_ACCOUNT_DATA_STATUS {
  SUCCESS = 'SUCCESS',
  INVALID_ACCESS_TOKEN = 'INVALID_ACCESS_TOKEN',
  INVALID_BUDGET_ID = 'INVALID_BUDGET_ID',
  GENERAL_ERROR = 'GENERAL_ERROR',
}

export interface YnabAccountDetails {
  budgets: BudgetSummary[];
  accounts: YnabFinancialAccount[];
  categories?: string[];
}

export interface YnabAccountDataType {
  ynabAccountData?: YnabAccountDetails;
  financialAccountDetails?: FinancialAccountDetails[];
  status: FETCH_YNAB_ACCOUNT_DATA_STATUS;
}
