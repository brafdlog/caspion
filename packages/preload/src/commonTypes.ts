// TODO: not all the types are needed here
import { type Auth } from 'googleapis';
import { type Transaction } from 'israeli-bank-scrapers-core/lib/transactions';
import { type BudgetSummary, type Account as YnabAccount } from 'ynab';

export type { ScraperScrapingResult } from 'israeli-bank-scrapers-core';

export type Credentials = Auth.Credentials;

export enum OutputVendorName {
  YNAB = 'ynab',
  GOOGLE_SHEETS = 'googleSheets',
  JSON = 'json',
  CSV = 'csv',
}

export interface Config {
  outputVendors: {
    [OutputVendorName.GOOGLE_SHEETS]?: GoogleSheetsConfig;
    [OutputVendorName.YNAB]?: YnabConfig;
    [OutputVendorName.JSON]?: JsonConfig;
    [OutputVendorName.CSV]?: CsvConfig;
  };
  scraping: {
    timeout: number;
    numDaysBack: number;
    showBrowser: boolean;
    accountsToScrape: AccountToScrapeConfig[];
    chromiumPath?: string;
    maxConcurrency?: number;
  };
}

export interface AppInfo {
  sourceCommitShort: string;
  repository: string;
  discordChanel: string;
  currentVersion: string;
}

export type OutputVendorConfigs = Exclude<Config['outputVendors'][OutputVendorName], undefined>;
export type OutputVendorConfig<T extends OutputVendorName> = Exclude<
  Config['outputVendors'][T],
  undefined
>;

interface OutputVendorConfigBase {
  active: boolean;
}

export interface CsvConfig extends OutputVendorConfigBase {
  options: {
    filePath: string;
  };
}

export interface JsonConfig extends OutputVendorConfigBase {
  options: {
    filePath: string;
  };
}

export interface GoogleSheetsConfig extends OutputVendorConfigBase {
  options: {
    credentials: Credentials;
    // credentials: string;
    spreadsheetId: string;
  };
}

export interface YnabConfig extends OutputVendorConfigBase {
  options: {
    accessToken: string;
    accountNumbersToYnabAccountIds: Record<string, string>;
    budgetId: string;
    maxPayeeNameLength?: number;
  };
}

export interface AccountToScrapeConfig {
  id: string;
  // key: CompanyTypes;
  key: string;
  name: string;
  loginFields: Record<string, string>;
  active?: boolean;
}

export enum CompanyTypes {
  HAPOALIM = 'hapoalim',
  HAPOALIM_BE_ONLINE = 'hapoalimBeOnline',
  BEINLEUMI = 'beinleumi',
  UNION = 'union',
  AMEX = 'amex',
  ISRACARD = 'isracard',
  VISACAL = 'visaCal',
  MAX = 'max',
  LEUMI_CARD = 'leumiCard',
  OTSAR_HAHAYAL = 'otsarHahayal',
  DISCOUNT = 'discount',
  MIZRAHI = 'mizrahi',
  LEUMI = 'leumi',
  MASSAD = 'massad',
  YAHAV = 'yahav',
  BEYAHAD_BISHVILHA = 'beyahadBishvilha',
  BEHATSDAA = 'behatsdaa',
  MERCANTILE = 'mercantile',
}

export enum AccountType {
  IMPORTER = 'Importer',
  EXPORTER = 'Exporter',
}

export enum AccountStatus {
  IDLE = 'idle',
  PENDING = 'pending',
  IN_PROGRESS = 'inProgress',
  DONE = 'done',
  ERROR = 'error',
}

export interface BudgetTrackingEvent {
  message: string;

  vendorId?: CompanyTypes | OutputVendorName;

  accountStatus?: AccountStatus;

  error?: Error;

  accountType?: AccountType;
}

/// ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// End of common types
/// ////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

export interface Log {
  message: string;
  originalEvent?: BudgetTrackingEvent;
}

export interface Account {
  id: string;
  companyId: string;
  displayName: string;
  logo: string;
  type: AccountType;
  active: boolean;
  status?: AccountStatus;
  logs: Log[];
}

export interface Importer extends Account {
  loginFields: Record<string, string>;
}

export interface Exporter extends Account {
  options: object;
}

export interface AccountMetadata {
  companyId: string;
  companyName: string;
  logo: string;
}

export enum ModalStatus {
  IMPORTER_SETTINGS,
  EXPORTER_SETTINGS,
  LOGS,
  HIDDEN,
  NEW_SCRAPER,
  GENERAL_SETTINGS,
}

export enum ExporterResultType {
  WEBSITE_URL,
  FILE,
}

export interface ExportResultMetadata {
  resultType: ExporterResultType;
  getResultUri(exporter: Exporter): string;
}

export type YnabFinancialAccount = Pick<YnabAccount, 'id' | 'name' | 'type'> & {
  budgetId: string;
  active: boolean;
};

export interface YnabAccountDetails {
  budgets: BudgetSummary[];
  accounts: YnabFinancialAccount[];
  categories?: string[];
}

export interface FinancialAccountDetails {
  name: string;
  accountNumber: string;
}

export type HandleScrapingEvent = (
  eventName: string,
  budgetTrackingEvent?: BudgetTrackingEvent,
) => void;

export enum FETCH_YNAB_ACCOUNT_DATA_STATUS {
  SUCCESS = 'SUCCESS',
  INVALID_ACCESS_TOKEN = 'INVALID_ACCESS_TOKEN',
  INVALID_BUDGET_ID = 'INVALID_BUDGET_ID',
  GENERAL_ERROR = 'GENERAL_ERROR',
}

export interface YnabAccountDataType {
  ynabAccountData?: YnabAccountDetails;
  financialAccountDetails?: FinancialAccountDetails[];
  status: FETCH_YNAB_ACCOUNT_DATA_STATUS;
}

export interface EnrichedTransaction extends Transaction {
  accountNumber: string;
  category?: string;
  hash: string;
}

export interface Config {
  outputVendors: {
    [OutputVendorName.GOOGLE_SHEETS]?: GoogleSheetsConfig;
    [OutputVendorName.YNAB]?: YnabConfig;
    [OutputVendorName.JSON]?: JsonConfig;
    [OutputVendorName.CSV]?: CsvConfig;
  };
  scraping: {
    numDaysBack: number;
    showBrowser: boolean;
    accountsToScrape: AccountToScrapeConfig[];
    chromiumPath?: string;
    maxConcurrency?: number;
    timeout: number;
  };
  useReactUI?: boolean;
}
