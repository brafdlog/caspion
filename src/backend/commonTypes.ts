import { Auth } from 'googleapis';
import { CompanyTypes, ScraperCredentials } from 'israeli-bank-scrapers-core';
import { BudgetSummary, Account } from 'ynab';
import { Transaction } from 'israeli-bank-scrapers-core/lib/transactions';

export type { ScaperScrapingResult } from 'israeli-bank-scrapers-core';
export interface Config {
  outputVendors: {
    [OutputVendorName.GOOGLE_SHEETS]?: GoogleSheetsConfig;
    [OutputVendorName.YNAB]?: YnabConfig;
    [OutputVendorName.JSON]?: JsonConfig;
    [OutputVendorName.CSV]?: CsvConfig;
  };
  scraping: {
    numDaysBack: number;
    showBrowser: boolean;
    accountsToScrape: AccountToScrapeConfig[];
    chromiumPath?: string;
    maxConcurrency?: number;
    timeout: number;
  },
  useReactUI?: boolean
}

export enum OutputVendorName {
  YNAB = 'ynab',
  GOOGLE_SHEETS = 'googleSheets',
  JSON = 'json',
  CSV = 'csv'
}

export type OutputVendorConfigs = Exclude<Config['outputVendors'][OutputVendorName], undefined>
export type OutputVendorConfig<T extends OutputVendorName> = Exclude<Config['outputVendors'][T], undefined>

interface OutputVendorConfigBase {
  active: boolean;
}

export interface CsvConfig extends OutputVendorConfigBase {
  options: {
    filePath: string;
  }
}

export interface JsonConfig extends OutputVendorConfigBase {
  options: {
    filePath: string;
  }
}

export interface GoogleSheetsConfig extends OutputVendorConfigBase {
  options: {
    credentials: Credentials;
    spreadsheetId: string;
  }
}

export interface YnabConfig extends OutputVendorConfigBase {
  options: {
    accessToken: string;
    accountNumbersToYnabAccountIds: { [key: string]: string };
    budgetId: string;
    maxPayeeNameLength?: number;
  };
}

export interface AccountToScrapeConfig {
  id: string;
  key: CompanyTypes;
  name: string;
  loginFields: ScraperCredentials;
  active?: boolean;
}

export interface EnrichedTransaction extends Transaction {
  accountNumber: string;
  category?: string;
  hash: string;
}

export type ExportTransactionsParams = {
  transactionsToCreate: EnrichedTransaction[];
  startDate: Date;
  outputVendorsConfig: Config['outputVendors'];
}

export type ExportTransactionsResult = {
  exportedTransactionsNum: number
}

export type ExportTransactionsFunction = (
  exportTransactionsParams: ExportTransactionsParams,
  eventPublisher: any
) => Promise<ExportTransactionsResult>;

export interface OutputVendor {
  name: OutputVendorName;
  init?: (outputVendorsConfig: Config['outputVendors']) => Promise<void>;
  exportTransactions: ExportTransactionsFunction;
}

export type OAuth2Client = Auth.OAuth2Client;
export type Credentials = Auth.Credentials;

export interface FinancialAccountDetails {
  name: string;
  accountNumber: string;
}

export type YnabFinancialAccount = Pick<Account, 'id' | 'name' | 'type'> & { budgetId: string; active: boolean }

export enum FETCH_YNAB_ACCOUNT_DATA_STATUS {
  SUCCESS = 'SUCCESS',
  INVALID_ACCESS_TOKEN = 'INVALID_ACCESS_TOKEN',
  INVALID_BUDGET_ID = 'INVALID_BUDGET_ID',
  GENERAL_ERROR = 'GENERAL_ERROR'
}

export interface YnabAccountDetails {
  budgets: BudgetSummary[];
  accounts: YnabFinancialAccount[];
  categories?: string[]
}

export type YnabAccountDataType = {
  ynabAccountData?: YnabAccountDetails,
  financialAccountDetails?: FinancialAccountDetails[],
  status: FETCH_YNAB_ACCOUNT_DATA_STATUS
}
