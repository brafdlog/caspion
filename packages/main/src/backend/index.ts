import { type Config } from '@/backend/commonTypes';
import { createTransactionsInExternalVendors } from '@/backend/export/exportTransactions';
import { scrapeFinancialAccountsAndFetchTransactions } from '@/backend/import/importTransactions';
import moment from 'moment';
import * as configManager from './configManager/configManager';
import * as Events from './eventEmitters/EventEmitter';
import outputVendors from './export/outputVendors';
import * as bankScraper from './import/bankScraper';

export { CompanyTypes } from 'israeli-bank-scrapers-core';
export { Events, configManager, outputVendors };

export const { inputVendors } = bankScraper;

export async function scrapeAndUpdateOutputVendors(
  config: Config,
  optionalEventPublisher?: Events.EventPublisher,
) {
  const eventPublisher = optionalEventPublisher ?? new Events.BudgetTrackingEventEmitter();

  const startDate = moment().subtract(config.scraping.numDaysBack, 'days').startOf('day').toDate();

  await eventPublisher.emit(Events.EventNames.IMPORT_PROCESS_START, {
    message: `Starting to scrape from ${startDate} to today`,
  });

  const companyIdToTransactions = await scrapeFinancialAccountsAndFetchTransactions(
    config.scraping,
    startDate,
    eventPublisher,
  );
  try {
    const executionResult = await createTransactionsInExternalVendors(
      config.outputVendors,
      companyIdToTransactions,
      startDate,
      eventPublisher,
    );

    return executionResult;
  } catch (e) {
    await eventPublisher.emit(
      Events.EventNames.GENERAL_ERROR,
      new Events.BudgetTrackingEvent({ message: (e as Error).message, error: e as Error }),
    );
    throw e;
  }
}
