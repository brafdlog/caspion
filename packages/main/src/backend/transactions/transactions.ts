import { type Transaction } from 'israeli-bank-scrapers-core/lib/transactions';
import _ from 'lodash';
import { type EnrichedTransaction } from '../commonTypes';
import { compareObjectsByDate } from './dates';

const { uniq } = _;

const unifyHash = (hash: string) =>
  hash
    .replace(/`/g, "'")
    .replace(/00\dZ/, '000Z') // Leumi: Last part of the date is sometimes 000Z, 002Z, 003Z...
    // eslint-disable-next-line no-control-regex
    .replace(/[\u0000-\u001F\u007F-\u009F\u200E]/g, '') // Special characters
    .replace('‏', ''); // Special character

const transactionArrayToUnifyHash = (transactions: EnrichedTransaction[]) =>
  transactions.reduce(
    (acc, enrichedTransaction) => {
      acc[unifyHash(enrichedTransaction.hash)] = enrichedTransaction;
      return acc;
    },
    {} as Record<string, EnrichedTransaction>,
  );

export const calculateTransactionHash = (
  { date, chargedAmount, description, memo }: Transaction,
  companyId: string,
  accountNumber: string,
) => {
  return unifyHash(`${date}_${chargedAmount}_${description}_${memo}_${companyId}_${accountNumber}`);
};

export const mergeTransactions = (a: EnrichedTransaction[], b: EnrichedTransaction[]) => {
  const aObj = transactionArrayToUnifyHash(a);
  const bObj = transactionArrayToUnifyHash(b);
  const hashes = uniq(Object.keys(aObj).concat(...Object.keys(bObj)));

  const mergedObj = hashes.reduce((merged: Record<string, EnrichedTransaction>, hash: string) => {
    merged[hash] = { ...aObj[hash], ...bObj[hash] };
    return merged;
  }, {});
  return Object.values(mergedObj);
};

export const filterExistedHashes = (
  transactions: EnrichedTransaction[],
  existingHashes: string[],
) => {
  const unifiedExistingHashs = existingHashes.map(unifyHash);
  return transactions.filter(({ hash }) => !unifiedExistingHashs.includes(unifyHash(hash)));
};

export const sortByDate = (transactions: EnrichedTransaction[]) =>
  transactions.sort(compareObjectsByDate);
