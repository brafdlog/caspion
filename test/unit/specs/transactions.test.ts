import { TransactionStatuses, TransactionTypes, Transaction } from 'israeli-bank-scrapers-core/lib/transactions';
import { calculateTransactionHash, mergeTransactions } from '../../../src/backend/transactions/transactions';

import { EnrichedTransaction } from '../../../src/backend/commonTypes';

const TRANSACTION_1 = {
  type: TransactionTypes.Normal,
  date: '2020-03-24T22:00:00.000Z',
  processedDate: '2020-06-01T21:00:00.000Z',
  originalAmount: -14.95,
  originalCurrency: 'USD',
  chargedAmount: -54.72,
  description: 'Puppies',
  memo: '',
  installments: undefined,
  status: TransactionStatuses.Completed
};
const TRANSACTION_1_HASH = '2020-03-24T22:00:00.000Z_-54.72_Puppies__visaCal_2222';

const TRANSACTION_2 = {
  type: TransactionTypes.Normal,
  date: '2020-03-27T21:00:00.000Z',
  processedDate: '2020-04-01T21:00:00.000Z',
  originalAmount: -200,
  originalCurrency: 'ILS',
  chargedAmount: -200,
  description: "מי גבעתיים בע'מ",
  memo: 'some memo',
  installments: undefined,
  status: TransactionStatuses.Completed
};

const TRANSACTION_2_HASH = "2020-03-27T21:00:00.000Z_-200_מי גבעתיים בע'מ_some memo_visaCal_2222";

const companyId = 'visaCal';
const accountNumber = '2222';

const enrich = (trx: Transaction): EnrichedTransaction => ({
  ...trx,
  accountNumber,
  hash: calculateTransactionHash(trx, companyId, accountNumber)
});

describe('Transactions', () => {
  test.each([
    [TRANSACTION_1, TRANSACTION_1_HASH],
    [TRANSACTION_2, TRANSACTION_2_HASH],
    [{ ...TRANSACTION_1, processedDate: '2020-06-01T21:00:00.002Z' }, TRANSACTION_1_HASH],
    [{ ...TRANSACTION_1, date: '2020-03-24T22:00:00.002Z' }, TRANSACTION_1_HASH],
    [{ ...TRANSACTION_2, description: 'מי גבעתיים בע`מ' }, TRANSACTION_2_HASH],
    [{ ...TRANSACTION_2, description: "מי גבעתיים\u200E בע'מ" }, TRANSACTION_2_HASH],
    [{ ...TRANSACTION_2, description: "‏מי גבעתיים בע'מ" }, TRANSACTION_2_HASH]
  ])('Transaction hash', (transaction, expectedHash) => {
    expect(calculateTransactionHash(transaction, companyId, accountNumber)).toEqual(expectedHash);
  });

  test('Merge Transactions', () => {
    const a = [TRANSACTION_1, TRANSACTION_2].map(enrich);
    const b = [
      { ...TRANSACTION_1, processedDate: '2020-06-01T21:00:00.002Z' },
      { ...TRANSACTION_1, date: '2020-03-24T22:00:00.002Z' },
      { ...TRANSACTION_2, description: 'מי גבעתיים בע`מ' },
      { ...TRANSACTION_2, description: "מי גבעתיים\u200E בע'מ" }
    ].map(enrich);

    const actualMerged: Transaction[] = mergeTransactions(a, b);

    expect(actualMerged.length).toBe(2);
    expect(actualMerged).toMatchSnapshot();
  });
});
