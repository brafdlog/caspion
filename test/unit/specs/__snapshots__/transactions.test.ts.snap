// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transactions Merge Transactions 1`] = `
Array [
  Object {
    "accountNumber": "2222",
    "chargedAmount": -54.72,
    "date": "2020-03-24T22:00:00.002Z",
    "description": "Puppies",
    "hash": "2020-03-24T22:00:00.000Z_-54.72_Puppies__visaCal_2222",
    "installments": undefined,
    "memo": "",
    "originalAmount": -14.95,
    "originalCurrency": "USD",
    "processedDate": "2020-06-01T21:00:00.000Z",
    "status": "completed",
    "type": "normal",
  },
  Object {
    "accountNumber": "2222",
    "chargedAmount": -200,
    "date": "2020-03-27T21:00:00.000Z",
    "description": "מי גבעתיים‎ בע'מ",
    "hash": "2020-03-27T21:00:00.000Z_-200_מיגבעתייםבע'מ_somememo_visaCal_2222",
    "installments": undefined,
    "memo": "some memo",
    "originalAmount": -200,
    "originalCurrency": "ILS",
    "processedDate": "2020-04-01T21:00:00.000Z",
    "status": "completed",
    "type": "normal",
  },
]
`;
